src/main/scala/org/neo4j/spark/service/SchemaService.scala

```diff
@@ -158,11 +168,178 @@ class SchemaService(private val options: Neo4jOptions, private val jobId: String
+        .getOrElse(Collections.emptyMap())
+        .asInstanceOf[util.Map[String, Long]].asScala
+      map.filterKeys(k => options.nodeMetadata.labels.contains(k))
+        .values.min
+    } else {
+      countForNodeWithQuery(filters)
+    }
+  } catch {
+    case e: ClientException =>
+      e.code match {
+        case "Neo.ClientError.Procedure.ProcedureNotFound" => {
+          countForNodeWithQuery(filters)
+        }
+        case _ => logExceptionForCount(e)
+      }
+    case e: Throwable => logExceptionForCount(e)
+  }
+
+  def countForRelationship(filters: Array[Filter]): Long = try {
+    if (filters.isEmpty) {
+      val query = "CALL apoc.meta.stats() yield relTypes RETURN relTypes"
+      val map = session.run(query).single()
+        .asMap()
+        .asScala
+        .get("relTypes")
+        .getOrElse(Collections.emptyMap())
+        .asInstanceOf[util.Map[String, Long]]
+        .asScala
+      val minFromSource = options.relationshipMetadata.source.labels
+        .map(_.quote())
+        .map(label => map.get(s"(:$label)-[:${options.relationshipMetadata.relationshipType}]->()").getOrElse(Long.MaxValue))
+        .min
+      val minFromTarget = options.relationshipMetadata.target.labels
+        .map(_.quote())
+        .map(label => map.get(s"()-[:${options.relationshipMetadata.relationshipType}]->(:$label)").getOrElse(Long.MaxValue))
+        .min
+      Math.min(minFromSource, minFromTarget)
+    } else {
+      countForRelationshipWithQuery(filters)
+    }
+  } catch {
+    case e: ClientException =>
+      e.code match {
+        case "Neo.ClientError.Procedure.ProcedureNotFound" => {
+          countForRelationshipWithQuery(filters)
+        }
+        case _ => logExceptionForCount(e)
+      }
+    case e: Throwable => logExceptionForCount(e)
+  }
+
+  private def logExceptionForCount(e: Throwable): Long = {
+    log.error("Cannot compute the count because the following exception:", e)
+    -1
+  }
+
+  def countForQuery(): Long = {
+    val queryCount: String = options.queryMetadata.queryCount
+    if (Neo4jUtil.isLong(queryCount)) {
+      queryCount.toLong
+    } else {
+      val query = if (queryCount.nonEmpty) {
+        options.queryMetadata.queryCount
+      } else {
+        s"""CALL { ${options.query.value} }
+           |RETURN count(*) AS count
+           |""".stripMargin
+      }
+      session.run(query).single().get("count").asLong()
+    }
+  }
+
+  def count(filters: Array[Filter] = Array.empty[Filter]): Long = options.query.queryType match {
+    case QueryType.LABELS => countForNode(filters)
+    case QueryType.RELATIONSHIP => countForRelationship(filters)
+    case QueryType.QUERY => countForQuery()
+  }
+
+  def skipLimitFromPartition(): Seq[(Long, Long)] = if (options.partitions == 1) {
+    Seq.empty
+  } else {
+    val count: Long = this.count()
+    if (count <= 0) {
+      Seq.empty
+    } else {
+      val partitionSize = Math.ceil(count / options.partitions).toLong
```

Comment:
isn't this a whole number division? so `ceil` has no effect

Reply:
Great catch thank you!!!